contracts/src/token/erc721/extensions/consecutive.rs

```diff
@@ -0,0 +1,1054 @@
+//! Implementation of the ERC-2309 "Consecutive Transfer Extension" as defined
+//! in https://eips.ethereum.org/EIPS/eip-2309[ERC-2309].
+//!
+//! This extension allows the minting large batches of tokens, during
+//! contract construction only. For upgradeable contracts, this implies that
+//! batch minting is only available during proxy deployment, and not in
+//! subsequent upgrades. These batches are limited to 5000 tokens at a time by
+//! default to accommodate off-chain indexers.
+//!
+//! Using this extension removes the ability to mint single tokens during
+//! contract construction. This ability is regained after construction. During
+//! construction, only batch minting is allowed.
+//!
+//! IMPORTANT: Function [`Erc721Consecutive::_mint_consecutive`] is not suitable
+//! to be called from constructor. Because of stylus sdk limitation. Function
+//! [`Erc721Consecutive::_stop_mint_consecutive`] should be called to end
+//! consecutive mint of tokens. After that, minting a token
+//! with [`Erc721Consecutive::_mint`] will be possible.
+//!
+//! IMPORTANT: This extension does not call the [`Erc721::_update`] function for
+//! tokens minted in batch. Any logic added to this function through overrides
+//! will not be triggered when token are minted in batch. You may want to also
+//! override [`Erc721Consecutive::_increaseBalance`] or
+//! [`Erc721Consecutive::_mintConsecutive`] to account for these mints.
+//!
+//! IMPORTANT: When overriding [`Erc721Consecutive::_mintConsecutive`], be
+//! careful about call ordering. [`Erc721Consecutive::owner_of`] may return
+//! invalid values during the [`Erc721Consecutive::_mintConsecutive`]
+//! execution if the super call is not called first. To be safe, execute the
+//! super call before your custom logic.
+
+use alloc::vec;
+
+use alloy_primitives::{uint, Address, U128, U256};
+use alloy_sol_types::sol;
+use stylus_proc::{external, sol_storage, SolidityError};
+use stylus_sdk::{
+    abi::Bytes, call::MethodError, evm, msg, prelude::TopLevelStorage,
+};
+
+use crate::{
+    token::{
+        erc721,
+        erc721::{
+            Approval, ERC721IncorrectOwner, ERC721InvalidApprover,
+            ERC721InvalidReceiver, ERC721InvalidSender, ERC721NonexistentToken,
+            Erc721, IErc721, Transfer,
+        },
+    },
+    utils::{
+        math::storage::{AddAssignUnchecked, SubAssignUnchecked},
+        structs::{
+            bitmap::BitMap,
+            checkpoints,
+            checkpoints::{Trace160, U96},
+        },
+    },
+};
+
+sol_storage! {
+    /// State of an [`Erc72Erc721Consecutive`] token.
+    #[cfg_attr(all(test, feature = "std"), derive(motsu::DefaultStorageLayout))]
+    pub struct Erc721Consecutive {
+        /// Erc721 contract storage.
```

Comment:
I am not sure if we should put each contract(struct) name in the docs in `[]` or not, like `[Erc721]`

Reply:
Do you think it is better to put "Erc721" into [`Erc721`]?

Reply:
I think that for struct fields it is fine if we don't